### PR TITLE
6914801: IPv6 unavailable if stdin is a socket

### PR DESCRIPTION
**Pull request by djelinski**
This patch reenables IPv6 stack when stdin is an IPv4 socket.

The code that blocked IPv6 was introduced in JDK-4673940 back when JDK could only operate with either IPv4 or IPv6 sockets, and was using `IPv6_available` to determine which socket type was in use. Now that JDK is able to operate with both IPv4 and IPv6 sockets at the same time, the check for IPv4 stdin is no longer relevant.

Included test passes with the changes applied, fails without them. Other tier1-3 tests also pass.

---------
### Progress
- [x] Change must be properly reviewed (1 review required, with at least 1 [Reviewer](https://openjdk.org/bylaws#reviewer))
- [x] Change must not contain extraneous whitespace
- [x] Commit message must refer to an issue

### Issue
 * [JDK-6914801](https://bugs.openjdk.org/browse/JDK-6914801): IPv6 unavailable if stdin is a socket

### Reviewers
 * [Michael McMahon](https://openjdk.org/census#michaelm) (@Michael-Mc-Mahon - **Reviewer**)

### Reviewing
<details><summary>Using <code>git</code></summary>

Checkout this PR locally: \
`$ git fetch https://git.openjdk.org/jdk pull/11638/head:pull/11638` \
`$ git checkout pull/11638`

Update a local copy of the PR: \
`$ git checkout pull/11638` \
`$ git pull https://git.openjdk.org/jdk pull/11638/head`

</details>
<details><summary>Using Skara CLI tools</summary>

Checkout this PR locally: \
`$ git pr checkout 11638`

View PR using the GUI difftool: \
`$ git pr show -t 11638`

</details>
<details><summary>Using diff file</summary>

Download this PR as a diff file: \
<a href="https://git.openjdk.org/jdk/pull/11638.diff">https://git.openjdk.org/jdk/pull/11638.diff</a>

</details>

### CODE REVIEW

**Comment by AlanBateman** [comment]
I don't have time to study the tests right now but the src change look okay.

**Comment by djelinski** [comment]
Happy new year!
Thanks Alan for checking this. Can I get a more formal review?

The test I added is basically a copy of EchoTest, except that 1) it runs only if an IPv6 socket can be opened by the harness, 2) the started application works only if it can open an IPv6 socket.
I also modified the tests to use IPv4 sockets for communication when they are available. If IPv4 is not available, the test will use platform-default socket type, and should still pass.

**Comment by Michael-Mc-Mahon** [comment]
> Happy new year! Thanks Alan for checking this. Can I get a more formal review?
> 
> The test I added is basically a copy of EchoTest, except that 1) it runs only if an IPv6 socket can be opened by the harness, 2) the started application works only if it can open an IPv6 socket. I also modified the tests to use IPv4 sockets for communication when they are available. If IPv4 is not available, the test will use platform-default socket type, and should still pass.

Checking it now.

File: test/jdk/java/nio/channels/spi/SelectorProvider/inheritedChannel/Launcher.java [hunk 1]
```diff
@@ -100,11 +101,15 @@ public static SocketChannel launchWithInetSocketChannel(String className,
...
-            SocketChannel sc1 = SocketChannel.open(isa);
+        ServerSocketChannel ch;
+        try {
+            ch = ServerSocketChannel.open(INET);
+        } catch (Exception e) {
+            ch = ServerSocketChannel.open();
+        }
+        try (ServerSocketChannel ssc = ch) {
+            ssc.socket().bind(new InetSocketAddress(InetAddress.getLoopbackAddress(), 0));
+            SocketChannel sc1 = SocketChannel.open(ssc.getLocalAddress());
```
Comment (Michael-Mc-Mahon) [inline test/jdk/java/nio/channels/spi/SelectorProvider/inheritedChannel/Launcher.java:112]: Could we add some logging/printlns here to indicate the protocol family of `ch` or its local address, to aid in debugging if we ever have problems with the test (eg on IPv6 only systems)?

Comment (djelinski) [reply test/jdk/java/nio/channels/spi/SelectorProvider/inheritedChannel/Launcher.java:115]: Added; now it produces output like:
```
Using INET (IPv4) channel
Socket bound to /127.0.0.1:44179
```

**Review by Michael-Mc-Mahon — APPROVED** [review]
LGTM

**Comment by AlanBateman** [comment]
> Happy new year! Thanks Alan for checking this. Can I get a more formal review?

I didn't get time to study the test changes but I see Michael did, so I think you are good to go.

**Pull request closed without merging**